### PR TITLE
KAFKA-3013

### PR DESCRIPTION
**Pull request by MayureshGharat**
Added topic-partition information to the exception message on batch expiry in RecordAccumulator

### CODE REVIEW

**Comment by gwenshap** [comment]
Thanks, this is super helpful!

Minor improvement: The error message should probably mention that the batch expired due to timeout while requesting metadata from brokers? i.e. lets not make our users read the code comment to figure out what happened.

**Comment by MayureshGharat** [comment]
I was just thinking that "requesting metadata" might not be the only reason. Suppose we have inflight request set to 1 and somehow the connection between the leader and the producer has some issues (cut cable), but the broker is able to communicate with other brokers in the cluster, so the metadata refresh still returns this broker as the leader. 
The batches sent on wire will be timed out and retried until the max retries. In the mean time, the other batches in the RecordAccumulator might expire due to timeout.

**Comment by darionyaphet** [comment]
Could add `recordCount` on TimeoutException Message to tell me how many records write failed in this batch ?  thank you :)

**Comment by lindong28** [comment]
@MayureshGharat I am a little confused by your example. Suppose the leader can communicate with other brokers, and other brokers can communicate with producer so that they can provide metadata, then in an IP-based network, the leader should also communicate with the producer, right?

**Comment by MayureshGharat** [comment]
Discussed with @lindong28 offline. We discussed the case where its is possible that the entire cluster is healthy but somehow the connection between the producer and a single broker is messed up.

**Comment by lindong28** [comment]
@MayureshGharat  Yeah I can not rule out the possibility. But I can not come up with a reasonable scenario that this may happen as well.. It seems an unlikely case if not impossible. Just my opinion.

**Comment by ijuma** [comment]
What was the conclusion with regards to the message? As it is, this PR is already an improvement, so it would be good to include it. If we can be a bit more specific, even better.

**Comment by MayureshGharat** [comment]
@ijuma do you mean we should include the "requesting metadata" part in the message?

**Comment by ijuma** [comment]
@MayureshGharat Yes, I was asking if there was a conclusion on whether we should include that part or not. It would be good to get @gwenshap's input since she asked for it and she would be able to merge this if she's happy with it. :)

**Comment by MayureshGharat** [comment]
@ijuma sure, this patch is waiting on @gwenshap's input, if she agrees with the explanation above. I am happy to submit another patch if required.

**Comment by MayureshGharat** [comment]
@gwenshap : Can you take another look at the patch and the reasoning in this PR?

**Comment by gwenshap** [comment]
I disagree that we can run into this issue when metadata is available. 

Note that we are checking request timeout against lastAttemptMs of the record batch. So as long as we are retrying, we shouldn't hit this type of timeout and expire.

If you want to validate: Set up IPFilters block between a producer and a single node in the cluster and check whether you are hitting the "expired" error.

Does that make sense?

**Comment by MayureshGharat** [comment]
Cool. I will update the PR. :)

**Comment by guozhangwang** [comment]
@gwenshap Could you take another look?

**Comment by ijuma** [comment]
LGTM (well, I actually suggested a tiny code clean-up as well)

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ public boolean maybeExpire(int requestTimeout, long now, long lingerMs) {
         if ((this.records.isFull() && requestTimeout < (now - this.lastAppendTime)) || requestTimeout < (now - (this.lastAttemptMs + lingerMs))) {
             expire = true;
             this.records.close();
-            this.done(-1L, new TimeoutException("Batch Expired"));
+            this.done(-1L, new TimeoutException("Batch containing " + recordCount + " record(s) expired due to timeout while requesting metadata from brokers for " + topicPartition.topic() + "-" + topicPartition.partition()));
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java]: Tiny nitpick: `TopicPartition.toString` does what you are doing manually here, so you could simplify it by just saying:

``` java
"Batch containing " + recordCount + " record(s) expired due to timeout while requesting metadata from brokers for " + topicPartition
```

**Comment by ijuma** [comment]
@MayureshGharat, unfortunately this branch has conflicts and it needs to be merged/rebased against master. I also left a minor comment on the PR itself. If you address these, we can try and get it merged by a committer.

**Comment by ijuma** [comment]
@gwenshap Can you please merge this?

**Comment by gwenshap** [comment]
Done. Sorry for the delay :)

**Comment by MayureshGharat** [comment]
NP. Thanks :)

-Mayuresh

On Mon, Mar 14, 2016 at 7:23 PM, Gwen (Chen) Shapira <
notifications@github.com> wrote:

> Done. Sorry for the delay :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/apache/kafka/pull/695#issuecomment-196619473

## 

-Regards,
Mayuresh R. Gharat
(862) 250-7125

**Pull request closed without merging**